File: Linked_Lists/206_reverse_linked_list/206.py
import sys
sys.path.append(r'/home/mount/Linked_Lists/')
from list import Single_Linked_List


def reverseList(head):
        
        if head == None:
            return head
        if head.next == None:
            return head
        
        last_node = reverseList(head.next)
        head.next.next = head
        head.next = None
        return last_node

linked_list = Single_Linked_List([1, 2, 3, 4, 5, 6])
linked_list.print()

root = linked_list.root
root = reverseList(root)
linked_list.print_from("result : ", root)

File: Linked_Lists/206_reverse_linked_list/206.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * T: O(N)
 * S: O(1)
 **/
// class Solution {
//     public ListNode reverseList(ListNode head) {

//         if (head == null) return head;

//         ListNode prev = null;

//         while (head != null) {
//             ListNode next = head.next;
//             head.next = prev;

//             prev = head;
//             head = next;
//         }

//         return prev;
//     }
// }

/**
 * T: O(N)
 * S: O(N)
 **/
class Solution {
    public ListNode reverseList(ListNode head) {

        if (head == null || head.next == null) return head;

        ListNode lastItem = reverseList(head.next);

        head.next.next = head;
        head.next = null;

        return lastItem;
    }
}